compiler/next/include/chpl/queries/Context-detail.h

```diff
@@ -64,73 +65,37 @@ template<typename ResultType, typename... ArgTs> class QueryMapResult;
 class QueryMapBase;
 template<typename ResultType, typename... ArgTs> class QueryMap;
 
-// define a hash function for std::tuple since the standard doesn't
-// include one.
-template<typename T>
-static std::size_t queryArgsHashOne(const T& v) {
-  std::hash<T> hasher;
-  return hasher(v);
-}
-
-template<size_t I, typename... Ts>
-static std::size_t queryArgsHashImpl(const std::tuple<Ts...>& tuple) {
-  size_t myhash = queryArgsHashOne(std::get<I>(tuple));
-  if (I + 1 < sizeof...(Ts)) {
-    size_t resthash =
-      queryArgsHashImpl<(I + 1 < sizeof... (Ts) ? I + 1 : I)>(tuple);
-    return chpl::hash_combine(myhash, resthash);
-  } else {
-    return myhash;
-  }
+template<typename TUP, size_t... I>
+static inline bool queryArgsEqualsImpl(const TUP& lhs, const TUP& rhs, std::index_sequence<I...>)
+{
+  // TODO: C++17 fold expression
+  // return(... && (std::get<I>(lhs) == std::get<I>(rhs)));
+
+  // C++14 code
+  // This expands the ret = ret && std::get<I>(lhs) == std::get<I>(rhs)
+  // assignments into a brace initializer list, evaluating the elements of
+  // I in left-to-right order. The value of ret is true iff all elements
+  // compare equal. The compiler optimizes the short-circuiting and
+  // sometimes generates SSE instructions to compare multiple words at once.
+  // The dummy variable is optimized away, as it is unused.
+  bool ret = true;
+  auto dummy = {(ret = ret && std::get<I>(lhs) == std::get<I>(rhs))...};
```

Comment:
Generally, :+1: here.

The previous version used `std::equal_to` because I imagine we might consider, for some type, defining `std::equal_to` to be different from `==`. However I am OK with this change for the time being because we don't do that yet.

In particular both the old compiler code and the new code define `std::less` for AST pointers -- here is the version in the old compiler:

https://github.com/chapel-lang/chapel/blob/ed08bc70765516e86704d2abb2899f310a41626c/compiler/include/baseAST.h#L438-L451

and here is the version in the new:

https://github.com/chapel-lang/chapel/blob/ed08bc70765516e86704d2abb2899f310a41626c/compiler/next/include/chpl/uast/ASTNode.h#L205-L225

The reason to do this is so that a `std::map` of AST node pointers will have deterministic ordering. Of course perhaps there is a better way to do this.

Reply:
It's very easy to change it to use `std::equal_to`.

Reply:
I added a small lambda helper-function `eq` to provide a generic interface for it to use, since it would become cumbersome to do it all in the expansion pack expression:
```
template<typename TUP, size_t... I>
static inline bool queryArgsEqualsImpl(const TUP& lhs, const TUP& rhs, std::index_sequence<I...>)
{
  // Lambda wrapper to std::equal_to                                                                                             
  auto eq = [](const auto& x, const auto& y) {
    return std::equal_to<std::decay_t<decltype(x)>>{}(x,y);
  };

  // TODO: C++17 fold expression                                                                                                 
  // return(... && eq(std::get<I>(lhs), std::get<I>(rhs)));                                                                      

  // C++14 code                                                                                                                  
  // This expands the ret = ret && std::get<I>(lhs) == std::get<I>(rhs)                                                          
  // assignments into a brace initializer list, evaluating the elements of                                                       
  // I in left-to-right order. The value of ret is true iff all elements                                                         
  // compare equal. The compiler optimizes the short-circuiting and                                                              
  // sometimes generates SSE instructions to compare multiple words at once.                                                     
  // The dummy variable is optimized away, as it is unused.                                                                      
  bool ret = true;
  auto dummy = {(ret = ret && eq(std::get<I>(lhs), std::get<I>(rhs)))...};
  return ret;
}
```
In C++14 and later, `auto` in a lambda function parameter acts like a templated type parameter, like `template<typename T>`.

In the code above, in order to call `std::equal_to<>`, we must provide a template argument for the type being compared. That would be very hard to do in the `{(ret = ret && std::get<I>(lhs) == std::get<I>(rhs))...}` expression if `==` were replaced with `std::equal_to<>()`.

The lambda is generic, and accepts any types for `x` and `y`; however, because both `x` and `y` come from the same element positions of the same tuple type, and the lambda is local to this function, `x` and `y` are always the same type. We use `std::decay_t<decltype(x)>` to get the type of the tuple element with cv and references removed (it would produce a compilation error if `y` were not compatible with this type).

Reply:
It looks like at the very least you'll need to update your explainer comment since it still uses `==` and put some of your explanation in for the lambda. However I think we can do better.

In my view, this:

``` c++

// define an equality function for std::tuple. the standard will call ==
// but we want to always compare our keys with equal_to
template<typename T>
static bool queryArgsEqualsOne(const T& lhs, const T& rhs) {
  std::equal_to<T> eq;
  return eq(lhs, rhs);
}
```

is an order of magnitude clearer than the lambda you have added here. Additionally it requires fewer C++ concepts required to understand it. (You may have figured out by now that I think it's really important to keep in mind that most C++ developers - including compiler developers - will not understand all the features of C++, never mind the new ones. So I think we should use them only when we really need to ).

So, I would recommend we use something along the lines of `queryArgsEqualsOne`. I think this also more similar to what we have ended up with `hash` where `hash_tuple_impl` calls `hash` for one element which uses `std::hash<T>`.